### ISSUE TITLE
Rect annotations with non-stacked histogram series

### ISSUE THREAD

**Issue by nickofthyme** [elastic/elastic-charts#907]
**Is your feature request related to a problem? Please describe.**
Using `RectAnnotations` with multiple non-stacked bar series with histogram mode enabled, forces the series to be stacked to provide a better data visualization. With this approach there is no issue with the `RectAnnotations`, however disabling histogram mode allows bars to be unstacked but result in a strange annotation placement.

> Note: The `RectAnnotations` is 10 hours from start of the domain.

### non-stacked / enableHistogramMode = true
![image](https://user-images.githubusercontent.com/19007109/99714596-64721f00-2a6b-11eb-999f-c1ab9e022747.png)

### non-stacked / enableHistogramMode = false
![image](https://user-images.githubusercontent.com/19007109/99714698-879cce80-2a6b-11eb-996a-cbfd5fa00a4f.png)

**Describe the solution you'd like**
Allow non-stacked bars with histogram mode enabled. This would be _somewhat_ strange with the bars not taking the full width of the interval but the grouping of the bars could suggest each bar covers the full interval.

![image](https://user-images.githubusercontent.com/19007109/99713746-4bb53980-2a6a-11eb-83ca-2918f75a0238.png)

**Describe alternatives you've considered**
Use `RectAnnotations` for the full interval. This has some drawbacks including:

1. On a time scale showing `RectAnnotations` that cover the full interval, using a `LineAnnotation` to show the current time would show within the interval and could be confusing to the user.
1. The current time partial bucket annotations used in visualize and discover indicate a ratio of the partial bucket, but with the full bucket being displayed, this ratio would be gone.

cc: @markov00 & @monfera

**Comment by nickofthyme** [comment]
Talked with @monfera over zoom about this. Takeaways:

1. Bad to have non-stacked clustered bars on continuous style time scale.
1. Annotation could be more descriptive and include the percent missing of the bucket/bin.
1. Annotations could be replaced with note in axis tick to indicate partial bucket/bin. See this 👇 

![image](https://user-images.githubusercontent.com/19007109/99720854-f8e07f80-2a73-11eb-9838-db6591716a9e.png)

Also noted issue with alignment of scale to time bucket values see https://github.com/elastic/elastic-charts/issues/908.

**Comment by monfera** [comment]
Thanks @nickofthyme for the summary! Yes, my 2 cents, after chatting about the situation is this; may need product/vis discussion about needs etc.:

tl;dr (as it's linked to a big topic): we could reassess and hopefully avoid the need for the grey bars, using time bin label or even dashboard/filter level annotation. The grey bars represent something to be avoided, due to mismatch in meaning: the superimposition of a continuous time projection onto an ordinal time projection.

=========

Details:

1. Binning implies ordinal, rather than continuous time scale: once we discretize times, eg. into a daily raster, we automatically have a discrete (specifically, ordinal) time axis now, instead of a continuous time axis. Ie. a bar chart, histogram(*), or a line or area chart with discretized time values would imply an ordinal scale 
2. A continuous time scale is applicable only for showing events at specific times; for scatterplots and connected scatterplots where we're not binning into temporal bins; and for eventual needle plots and stripe charts
3. Even if we just discretize time, but don't aggregate, eg. snap arbitrary time stamps to hour-start bounds, it leads to an ordinal scale
4. A rule of thumb that follows from 1.-3. is that snapping time (usually, but not always, due to aggregation of values) into a time raster leads to an ordinal time scale, but if there's no discretization of time, eg. we have documents with time stamps, then it leads to a continuous time scale
5. It's problematic to project both continuous and ordinal time onto the same surface. So a histogram with daily bins should NOT be superimposed by a timestamped event that vertically cuts a histogram bar at the 50% width, indicating that the event happened at noon
6. As a consequence of all this is that it's not advised to indicate partial bins on the opening or closing of the reported time via a grey bar of specific width, or other visual method which implies (encodes) something via the width, so none of the three screenshots in the original description is recommended. Reason: width-encoding a grey shaded area would amount to a composition of an ordinal and a continuous time projection, which is best avoided

OK what can be done?
1. As on the screenshot, the temporal aspect of the time bin is already covered in the X axis bin label, with the usual meaning that each label stands for a fixed unit of time, eg. one day; if it's not true, it can be indicated in the X axis bin label (I didn't call it tick label, because ticks allude to continuous time)
2. Another legit strategy is to simply chop off partial bins—it's my favorite because it avoids a whole can of worms that are present for even the simplest aggregation; eg.
- count, sum will be more misleading
- even the suitable aggregations, eg. mean or median, are a problem, because there may be seasonality within the time bin;
- more complex, or more derived functions, eg. showing differences, moving averages etc. will hinge on usually questionable data
- as of now, I'm not aware of a solid requirement for showing partial bins
- if not showing partial bins is an issue for the user, we probably didn't offer an adequate resolution, and there are too few bars left; after all
3. As an alternative to 2., it's possible to extend to full bins, rather than chopping off partial bins, or do it one way on the left and another way on the right
4. OK but how to add timestamped, continuous-time annotation onto a time histogram? 
- Solution 1: small multiples; top chart has bars and lines, while bottom chart has event markers on a continuous scale; the time anchors and projection scale at the raster can be shared
- Solution 2: aggregate annotations themselves into the same temporal raster, which can now be shown on a shared, ordinal projection
- Solution 3: use a continuous scale; it doesn't prevent short, horizontal line sections that represent eg. hourly count, sum or average values (called reference lines in eg. Tableau); it's almost like a Kibana histogram (which is not a real histogram(*)) but isn't bar-like

One legit case for a partial bin on the right is when we have streaming data and we're partially in the current bin, eg. reporting the past several days in a daily raster, including the (as yet incomplete) today - in this case, there's no need for a grey bar, as it's taken for granted, and the view (entire dashboard) needs to be time stamped, eg. "as of 2:49pm"; in fact a grey bar would be just noise.

Charts are for visually reading patterns after all; they're not tabular, numeric reports. 

So it'd be best to go deeper into the need and challenge ongoing assumptions about the perceived need for grey bars or even, partial bins (again, except partially reported day, which is best addressed differently).

=========

(*) Alright, histograms, in theory, operate on the domain of the independent variable, ie. in case of time, a potentially continuous scale, with the possibility that bins are not equal widths; double the width of a single bar in a histogram implies double the variable extent. 

But our histograms aren't really histograms, because histograms serve the purpose of showing a distribution. What's called histogram in Kibana is usually called a time series bar chart. And it doesn't end there. For example, the line/area chart equivalent of the histogram is called [frequency polygons](https://www.ck12.org/statistics/frequency-polygons/lesson/Frequency-Polygons-BSC-PST/). In Kibana, are we calling it area chart, histogram or frequency polygons? It is fundamentally just a time series chart. It shows time evolution, not distribution.

That the word "histogram" is misleading becomes most apparent with the clustered (a.k.a. grouped) bar chart aesthetic. There are [stacked histograms](https://www.originlab.com/doc/Origin-Help/Stacked-Histograms-Graph) but they're a rarity (let's for now forget about the oddity that we don't use it as histogram, we use it as time series bar chart). But I'm not aware of clustered histograms. If there's such a need it's best to do them in small multiples, or as stacked bars, or via interactivity, eg. category filtering.

In general, clustered bar charts are a dying species, I think. They have numerous weaknesses: they're hard to read. We could consider switching to small multiples or stacked bars. The best extant use of clustered bar charts is in financial BI, for the case when there are two categories, typically, plan vs actual, target vs actual or forecast vs actual. And even there it's challenging, and folks innovate things that have various tradeoffs, eg. bullet charts or IBCS-style charts. So we could revisit the recommendation, default or even availability of clustered bar charts.
![image](https://user-images.githubusercontent.com/1548516/99732694-3965ea80-2ac0-11eb-8448-43715b0098d1.png)
